src/instance.rs

```diff
@@ -25,18 +25,130 @@ pub trait PyObjectWithGIL: Sized {
 #[doc(hidden)]
 pub trait PyNativeType: PyObjectWithGIL {}
 
+/// A reference of type `T` which is in Python heap.
+///
+/// We can't implement `ToPyPointer` or `ToPyObject` for `pyclass`es, because they're not python
+/// objects untill copied to Python heap. So, instead of treating `&pyclass` as a Python object, we
+/// need to use special reference types `PyRef` and `PyRefMut`.
+/// ```
+/// # extern crate pyo3; fn main() {
+/// use pyo3::prelude::*;
+/// use pyo3::types::IntoPyDict;
+/// #[pyclass]
+/// struct Point {
+///     x: i32,
+///     y: i32,
+/// }
+/// #[pymethods]
+/// impl Point {
+///     fn length(&self) -> i32 {
+///         self.x * self.y
+///     }
+/// }
+/// let gil = Python::acquire_gil();
+/// let py = gil.python();
+/// let obj = py.init_ref(|| Point { x: 3, y: 4 }).unwrap();
+/// let d = vec![("p", obj)].into_py_dict(py);
+/// py.run("assert p.length() == 12", None, Some(d)).unwrap();
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct PyRef<'a, T> {
+    inner: &'a T,
+}
+
+impl<'a, T> PyRef<'a, T> {
+    pub(crate) fn new(t: &'a T) -> Self {
+        PyRef { inner: t }
+    }
+}
+
+impl<'a, T: PyTypeInfo> ToPyPointer for PyRef<'a, T> {
+    fn as_ptr(&self) -> *mut ffi::PyObject {
+        unsafe { (self.inner as *const _ as *mut u8).offset(-T::OFFSET) as *mut _ }
+    }
+}
+
+impl<'a, T: PyTypeInfo> ToPyObject for PyRef<'a, T> {
+    fn to_object(&self, py: Python) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+    }
+}
+
+impl<'a, T> Deref for PyRef<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.inner
+    }
+}
+
+impl<'a, T> Into<&'a PyObjectRef> for PyRef<'a, T>
```

Comment:
https://doc.rust-lang.org/nightly/std/convert/trait.Into.html:

> Library authors should not directly implement this trait, but should prefer implementing the From trait, which offers greater flexibility and provides an equivalent Into implementation for free, thanks to a blanket implementation in the standard library.

Reply:
Thanks, I didn't know that